src/control/lib/support/log.go

```diff
@@ -659,85 +669,232 @@ func copyServerConfig(log logging.Logger, opts ...CollectLogsParams) error {
-			return err
-		}
-	} else {
-		var startTime, endTime string
+// Calculate the start/end time provided by user.
+func getDateTime(log logging.Logger, opts ...CollectLogsParams) (time.Time, time.Time, error) {
+	var startTime, endTime, dateTimeFormat string
+	currentTime := time.Now()
 
-		logStartDate, err := time.Parse(MMDD, opts[0].LogStartDate)
-		if err != nil {
-			return err
-		}
+	logStartDate, err := time.Parse(MMDD, opts[0].LogStartDate)
+	if err != nil {
+		return currentTime, currentTime, err
+	}
 
-		logEndDate, err := time.Parse(MMDD, opts[0].LogEndDate)
-		if err != nil {
-			return err
-		}
+	logEndDate, err := time.Parse(MMDD, opts[0].LogEndDate)
+	if err != nil {
+		return currentTime, currentTime, err
+	}
 
-		// Default Start time, in case no start time provides on start dates.This will copy log start of the day.
-		if opts[0].LogStartTime == "" {
-			opts[0].LogStartTime = "00:00:00"
-		}
+	// Default Start time, in case no start time provides on start dates.This will copy log start of the day.
+	if opts[0].LogStartTime == "" {
+		opts[0].LogStartTime = "00:00:00"
+	}
 
-		// Default End time, in case no End time provides.This will copy log till the End of the day.
-		if opts[0].LogEndTime == "" {
-			opts[0].LogEndTime = "23:59:59"
-		}
+	// Default End time, in case no End time provides.This will copy log till the End of the day.
+	if opts[0].LogEndTime == "" {
+		opts[0].LogEndTime = "23:59:59"
+	}
 
+	// Set the date/time format for Engine log.
+	if opts[0].LogCmd == "EngineLog" {
 		startTime = fmt.Sprintf("%d/%d-%s", logStartDate.Month(), logStartDate.Day(), opts[0].LogStartTime)
 		endTime = fmt.Sprintf("%d/%d-%s", logEndDate.Month(), logEndDate.Day(), opts[0].LogEndTime)
+		dateTimeFormat = MMDDHHMMSS
+	}
 
-		actStartTime, _ := time.Parse(MMDDHHMMSS, startTime)
-		actEndTime, _ := time.Parse(MMDDHHMMSS, endTime)
+	// Set the date/time format for Control and Helper log.
+	if opts[0].LogCmd == "ControlLog" || opts[0].LogCmd == "HelperLog" {
+		startTime = fmt.Sprintf("%d/%d %s", logStartDate.Month(), logStartDate.Day(), opts[0].LogStartTime)
+		endTime = fmt.Sprintf("%d/%d %s", logEndDate.Month(), logEndDate.Day(), opts[0].LogEndTime)
+		dateTimeFormat = MMDD_HHMMSS
+	}
 
-		// Create the new empty file, which will be used to copy the matching date/time log lines.
-		logFileName := filepath.Join(destFile, filepath.Base(srcFile))
-		writeFile, err := os.Create(logFileName)
-		if err != nil {
-			return err
-		}
-		defer writeFile.Close()
+	actStartTime, _ := time.Parse(dateTimeFormat, startTime)
+	actEndTime, _ := time.Parse(dateTimeFormat, endTime)
 
-		readFile, err := os.Open(srcFile)
-		if err != nil {
-			return err
-		}
-		defer readFile.Close()
+	return actStartTime, actEndTime, nil
+}
+
+// Copy only specific lines from the server logs based on the Start/End date and time, provided by user.
+func cpLinesFromLog(log logging.Logger, srcFile string, destFile string, opts ...CollectLogsParams) error {
```

Comment:
I'm struggling a little with understanding the implementation of this function. Can you refactor a little to reduce code repetition? I see the algorithm as something like:
1) Read line from src.
2) Harvest date from line.
3) If date in desired range, enable copying. Else disable copying.
4) If copying is enabled, write the line to dest.
5) Go back to step 1.

The only thing that seems really different between the logs is the date format. Maybe that format could be selected at the beginning based on the log type, so that all the other code could be shared without much repetition? Regular expressions are built into Go, seems like a good use case for them.

Reply:
ok I will look at this code and see if I can reduce the repetition.